### ISSUE TITLE
FIM module in auditbeat keeps too many file handles open on Kubrenetes 

### ISSUE THREAD

**Issue by premendrasingh** [elastic/beats#19056]
I'm using Auditbeat with FIM module on Kubernetes daemonset with 40 pods on it.

Auditbeat version - latest
OS - Debian GNU/Linux 9
ulimit -n 1048576
Auditbeat pod memory allocation - 200mb

Open file handles go up to 2700 over 9 hours, then auditbeat pod gets OOMKilled and restarts.
Any suggestions how to close file handles.

**Comment by vjsamuel** [comment]
@andrewkroh can you please help us understand if this due to lack of clean up or something?

**Comment by andrewkroh** [comment]
Could you share what the Auditbeat process has in its /proc/pid/fds directory. This way we can get an idea if it's inotify inodes or actual file handles.

Do you have any graphs showing the number of open FDs over time that you can share? Also Auditbeat logs this metric every 30s by default. I'm curious to know if the number is high from the start or if it slowing grows over time. Is autodiscover creating/stopping FIM modules a lot? Or is the number of modules mostly static after Auditbeat starts?

**Comment by premendrasingh** [comment]
@andrewkroh  I have collected the output of lsof and sorted it below

```
$ awk '{print $9;}' /tmp/auditbeat-c2dws.txt | sort | uniq -c | sort -nr
 452 pipe
 225 [eventpoll]
 224 inotify
   2 (path
```

Number grows over time, as shown in graph for last 12 hours.  How can I check whether "autodiscover creating/stopping FIM modules a lot"?

<img width="879" alt="FileHandleCounts" src="https://user-images.githubusercontent.com/7782889/84095363-f8679700-a9b3-11ea-98cc-2a433da307be.png">

**Comment by andrewkroh** [comment]
I think there might be a `libbeat.config.module.{running,starts,stops}` metric that you can look at. I hope those reflect changes when autodiscover starts/stops modules (didn't check this yet). Debug logs could also help you confirm when modules start.

**Comment by andrewkroh** [comment]
I tried reloading the FIM module many times on Linux to see if any leaks were happening. I used the `auditbeat.config.modules` reload feature. With that path it appeared that as the module was stopped it was releasing its resources.

This path is different that autodiscover, but it does give some confidence that file_integrity module is implemented correctly. I need to try out the autodiscover path and see if it closing modules.

Maybe its not leaking and just has lots of directories it's monitoring? When pods stop on the Auditbeat node do the number of open handles decrease?
